apps/hashdotai/glossary/verification.mdx

```diff
@@ -0,0 +1,66 @@
+- Two processes **keep retrying an action** that fails (e.g. because they are in competition for the same resource) instead of resolving a conflict.
+
+## Deadlocks
+
+A **deadlock** happens when a process **gets stuck indefinitely** with no activity, e.g.
+
+- if a process **requires manual approval** but no user completes it
+- because two or more tasks are **waiting for each other** to complete
+- technical deadlocks such as two tasks waiting on each other to release a lock on a database row
+
+Potential solutions include:
+
+- timeouts and escalation paths for user tasks
+- avoid mutual dependencies between tasks
+- strategies to avoid database deadlocks, e.g.
+  - use optimistic locking, or
+  - (if the system requires pessimistic locking) consistent ordering of lock acquisition (i.e. make sure that Task A and Task B will always attempt to lock the same rows in the same order, so they can’t end up in a situation where one locks X then waits for Y, and the other locks Y then waits for X)
+
+## Starvation
+
+**Resource starvation** happens when a transaction or process **waits indefinitely** because it **can’t get access to a resource** due to other higher-priority tasks always taking precedence.
+
+# Verification in HASH
+
+## Processes (Flows)
+
+Processes in HASH are represented as what mathematicians call "_Petri nets_". Unlike most representations of process models, Petri nets can be formally verified, which means we can check their integrity in a number of different ways, including by looking at properties such as:
+
+1. **Reachability**
+  - A state is considered _reachable_ if it can be achieved from an initial marking by firing a sequence of transitions.
+  - Helps determine if a process can get stuck or fail to reach a goal state.
+  - **Example:** Can an order-processing system always reach the "Order Completed" state?
+  - **Verification:** HASH constructs a _reachability graph_ and uses _state-space exploration_ to check if any given target state can be reached. If it cannot, the process map editor is notified.
+2. **Boundedness** (Safeness)
+  - A Petri net is _k-bounded_ if no place contains more than _k tokens_ at any time.
+  - A _1-bounded_ net (Safe Petri Net) ensures that places hold at most one token, preventing overflow issues.
+  - Helps prevent resource exhaustion in concurrent systems.
+  - **Verification:** HASH analyzes the _incidence matrix_ and _state-space exploration_ to check if any place can exceed k tokens.
+3. **Liveness**
+  - A Petri net is considered _live_ if all transitions remain potentially fireable (i.e. no part of the system permanently stops working).
+  - Ensures that no process is permanently blocked (i.e. there are no _deadlocks_).
+  - **Example:** In an order-processing workflow, the **"Dispatch Order"** transition should always be possible if stock is available. If not, there’s a deadlock.
```

Comment:
```suggestion
  - Ensures that no process is permanently blocked (e.g. there are no _deadlocks_).
  - **Example:** In an order-processing workflow, if all necessary conditions are satisfied (e.g., payment confirmed, stock available) but the **"Dispatch Order"** transition still never fires, there is a liveness problem. This could indicate a deadlock or some other form of blocking that prevents the system from progressing.
```